# Coursera IBM Capstone Projects: Locations in Toronto's Neighborhoods
## Capstone Assignment week 3 (scraping web data, applying location api, clustering, visualization)

### Stage 1: gather postal codes from Wikipedia 

In [1]:
# import the necessary libraries and setup pandas set_option
import requests
from bs4 import BeautifulSoup
import pandas as pd
pd.set_option('display.max_colwidth', 200)
pd.set_option('display.max_rows', None)
import numpy as np
# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

The postal code data frame needs the columns: 'PostalCode', 'Borough', 'Neighborhood'.
Postal codes beginning with 'M' belong to the city of Toronto, see: https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M

In [2]:
# get the html file
wiki_url = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
file = requests.get(wiki_url).text

In [3]:
# use Beautifulsoup to extract necessary information
soup = BeautifulSoup(file, 'html5lib')

# get wikipedia table, postal codes are stored in the first <tbody></tbody>
wiki_table = soup.find('tbody')

# extract text and append to a new list, create a dictionary with the data, and finally create a pandas.DataFrame from the dictionary
code_list = []
for td in wiki_table.findAll('td'):
    code_list.append(td.text.strip())

code_dic = {}
code_dic['PostalCode'] = code_list[0::3]
code_dic['Borough'] = code_list[1::3]
code_dic['Neighborhood'] = code_list[2::3]
    
postal_df = pd.DataFrame(code_dic, columns = ['PostalCode', 'Borough', 'Neighborhood'])
postal_df.shape

(288, 3)

In [4]:
# clean the data fram
# drop rows whose value is 'Not assigned' =  Ignore cells with a borough that is Not assigned.
postal_df = postal_df[postal_df.Borough != 'Not assigned']

# combine rows with same PostalCode and concatenate the values in Neighborhood
postal_df = postal_df.groupby(['PostalCode', 'Borough'], sort=True, as_index=False).agg(lambda x: ', '.join(x.dropna()))

# replace the Neighborhood column with Borough if Neighborhood is not assigned
postal_df['Neighborhood'] = np.where(postal_df['Neighborhood'] == 'Not assigned', postal_df['Borough'], postal_df['Neighborhood'])
postal_df

PostalCode           Borough  \
0          M1B       Scarborough   
1          M1C       Scarborough   
2          M1E       Scarborough   
3          M1G       Scarborough   
4          M1H       Scarborough   
5          M1J       Scarborough   
6          M1K       Scarborough   
7          M1L       Scarborough   
8          M1M       Scarborough   
9          M1N       Scarborough   
10         M1P       Scarborough   
11         M1R       Scarborough   
12         M1S       Scarborough   
13         M1T       Scarborough   
14         M1V       Scarborough   
15         M1W       Scarborough   
16         M1X       Scarborough   
17         M2H        North York   
18         M2J        North York   
19         M2K        North York   
20         M2L        North York   
21         M2M        North York   
22         M2N        North York   
23         M2P        North York   
24         M2R        North York   
25         M3A        North York   
26         M3B        North York   
27         M3C        North York   
28         M3H        North York   
29         M3J        North York   
30         M3K        North York   
31         M3L        North York   
32         M3M        North York   
33         M3N        North York   
34         M4A        North York   
35         M4B         East York   
36         M4C         East York   
37         M4E      East Toronto   
38         M4G         East York   
39         M4H         East York   
40         M4J         East York   
41         M4K      East Toronto   
42         M4L      East Toronto   
43         M4M      East Toronto   
44         M4N   Central Toronto   
45         M4P   Central Toronto   
46         M4R   Central Toronto   
47         M4S   Central Toronto   
48         M4T   Central Toronto   
49         M4V   Central Toronto   
50         M4W  Downtown Toronto   
51         M4X  Downtown Toronto   
52         M4Y  Downtown Toronto   
53         M5A  Downtown Toronto   
54         M5B  Downtown Toronto   
55         M5C  Downtown Toronto   
56         M5E  Downtown Toronto   
57         M5G  Downtown Toronto   
58         M5H  Downtown Toronto   
59         M5J  Downtown Toronto   
60         M5K  Downtown Toronto   
61         M5L  Downtown Toronto   
62         M5M        North York   
63         M5N   Central Toronto   
64         M5P   Central Toronto   
65         M5R   Central Toronto   
66         M5S  Downtown Toronto   
67         M5T  Downtown Toronto   
68         M5V  Downtown Toronto   
69         M5W  Downtown Toronto   
70         M5X  Downtown Toronto   
71         M6A        North York   
72         M6B        North York   
73         M6C              York   
74         M6E              York   
75         M6G  Downtown Toronto   
76         M6H      West Toronto   
77         M6J      West Toronto   
78         M6K      West Toronto   
79         M6L        North York   
80         M6M              York   
81         M6N              York   
82         M6P      West Toronto   
83         M6R      West Toronto   
84         M6S      West Toronto   
85         M7A      Queen's Park   
86         M7R       Mississauga   
87         M7Y      East Toronto   
88         M8V         Etobicoke   
89         M8W         Etobicoke   
90         M8X         Etobicoke   
91         M8Y         Etobicoke   
92         M8Z         Etobicoke   
93         M9A         Etobicoke   
94         M9B         Etobicoke   
95         M9C         Etobicoke   
96         M9L        North York   
97         M9M        North York   
98         M9N              York   
99         M9P         Etobicoke   
100        M9R         Etobicoke   
101        M9V         Etobicoke   
102        M9W         Etobicoke   

                                                                                                                               Neighborhood  
0                                                                                                                    

In [5]:
# print the shape of the data frame
postal_df.shape

(103, 3)

### Stage 2:  get the latitude and the longitude coordinates of each neighborhood

In [6]:
# the geocoder module is working badly, so use the csv file
# alternative: 
!wget -q -O 'geospatial.csv' https://cocl.us/Geospatial_data

In [7]:
# read file into pandas and check the first 5 lines
geo_df = pd.read_csv('geospatial.csv')
geo_df.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [8]:
# check the shape of the data frame
geo_df.shape

(103, 3)

The shape looks alright, since there are also 103 rows. But the column 'Postal Code' differs to 'PostalCode' in the previous dafa frame. 

In [9]:
# rename the column 'Postal Code' to 'PostalCode'
geo_df.rename(columns={'Postal Code':'PostalCode'}, inplace=True)

In [10]:
# merge postal_df and geo_df
postal_geo_df = pd.merge(postal_df, geo_df, on = 'PostalCode')
postal_geo_df

PostalCode           Borough  \
0          M1B       Scarborough   
1          M1C       Scarborough   
2          M1E       Scarborough   
3          M1G       Scarborough   
4          M1H       Scarborough   
5          M1J       Scarborough   
6          M1K       Scarborough   
7          M1L       Scarborough   
8          M1M       Scarborough   
9          M1N       Scarborough   
10         M1P       Scarborough   
11         M1R       Scarborough   
12         M1S       Scarborough   
13         M1T       Scarborough   
14         M1V       Scarborough   
15         M1W       Scarborough   
16         M1X       Scarborough   
17         M2H        North York   
18         M2J        North York   
19         M2K        North York   
20         M2L        North York   
21         M2M        North York   
22         M2N        North York   
23         M2P        North York   
24         M2R        North York   
25         M3A        North York   
26         M3B        North York   
27         M3C        North York   
28         M3H        North York   
29         M3J        North York   
30         M3K        North York   
31         M3L        North York   
32         M3M        North York   
33         M3N        North York   
34         M4A        North York   
35         M4B         East York   
36         M4C         East York   
37         M4E      East Toronto   
38         M4G         East York   
39         M4H         East York   
40         M4J         East York   
41         M4K      East Toronto   
42         M4L      East Toronto   
43         M4M      East Toronto   
44         M4N   Central Toronto   
45         M4P   Central Toronto   
46         M4R   Central Toronto   
47         M4S   Central Toronto   
48         M4T   Central Toronto   
49         M4V   Central Toronto   
50         M4W  Downtown Toronto   
51         M4X  Downtown Toronto   
52         M4Y  Downtown Toronto   
53         M5A  Downtown Toronto   
54         M5B  Downtown Toronto   
55         M5C  Downtown Toronto   
56         M5E  Downtown Toronto   
57         M5G  Downtown Toronto   
58         M5H  Downtown Toronto   
59         M5J  Downtown Toronto   
60         M5K  Downtown Toronto   
61         M5L  Downtown Toronto   
62         M5M        North York   
63         M5N   Central Toronto   
64         M5P   Central Toronto   
65         M5R   Central Toronto   
66         M5S  Downtown Toronto   
67         M5T  Downtown Toronto   
68         M5V  Downtown Toronto   
69         M5W  Downtown Toronto   
70         M5X  Downtown Toronto   
71         M6A        North York   
72         M6B        North York   
73         M6C              York   
74         M6E              York   
75         M6G  Downtown Toronto   
76         M6H      West Toronto   
77         M6J      West Toronto   
78         M6K      West Toronto   
79         M6L        North York   
80         M6M              York   
81         M6N              York   
82         M6P      West Toronto   
83         M6R      West Toronto   
84         M6S      West Toronto   
85         M7A      Queen's Park   
86         M7R       Mississauga   
87         M7Y      East Toronto   
88         M8V         Etobicoke   
89         M8W         Etobicoke   
90         M8X         Etobicoke   
91         M8Y         Etobicoke   
92         M8Z         Etobicoke   
93         M9A         Etobicoke   
94         M9B         Etobicoke   
95         M9C         Etobicoke   
96         M9L        North York   
97         M9M        North York   
98         M9N              York   
99         M9P         Etobicoke   
100        M9R         Etobicoke   
101        M9V         Etobicoke   
102        M9W         Etobicoke   

                                                                                                                               Neighborhood  \
0                                                                                                                   

### Stage 3: explore the locations in Toronto and cluster neighborhoods

In [11]:
# install necessary modules
! conda install -c conda-forge geopy --yes
! conda install -c conda-forge folium =0.5.0 --yes

Solving environment: done

# All requested packages already installed.


CondaValueError: invalid package specification: =0.5.0



In [12]:
# import necessary libraries
from geopy.geocoders import Nominatim
import folium
from sklearn.cluster import KMeans

first, plot city map of Toronto with its boroughs and neighborhoods

In [13]:
# get coordinates for Toronto using Nominatim from geopy.geocoders
address = 'Toronto, Ontario, Canada'
geolocater = Nominatim(user_agent='ycc')
location = geolocater.geocode(address)
latitude = location.latitude
longitude = location.longitude

# create map of Toronto with the boroughs and neighborhoods using folium
map_toronto = folium.Map(location = [latitude, longitude], zoom_start = 10)
for lat, lng, borough, neighborhood in zip(postal_geo_df['Latitude'], postal_geo_df['Longitude'], postal_geo_df['Borough'], postal_geo_df['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker((lat, lng), radius=5, popup=label, color='pink', fill=True, fill_color='#3186cc', fill_opacity=0.6, parse_html=False).add_to(map_toronto)

map_toronto    

Downtown Toronto seems to be the most crowded area. Explore venues in this area.<br>
I. --> get Downtown Toronto's neighborhoods and their coordinates,

In [14]:
# get Downtown Toronto data from the postal_geo_df
downtown_df = postal_geo_df[postal_geo_df['Borough'] == 'Downtown Toronto'].reset_index(drop=True)
downtown_df.head()

PostalCode           Borough                 Neighborhood   Latitude  \
0        M4W  Downtown Toronto                     Rosedale  43.679563   
1        M4X  Downtown Toronto  Cabbagetown, St. James Town  43.667967   
2        M4Y  Downtown Toronto         Church and Wellesley  43.665860   
3        M5A  Downtown Toronto    Harbourfront, Regent Park  43.654260   
4        M5B  Downtown Toronto     Ryerson, Garden District  43.657162   

   Longitude  
0 -79.377529  
1 -79.367675  
2 -79.383160  
3 -79.360636  
4 -79.378937

II. --> get Downtown Toronto's coortinates and plot the borough with its neighborhoods

In [15]:
# get Downtown Toronto coordinates using Nominatim again
address = 'Downtown Toronto, Toronto, Ontario, Canada'
geolocater = Nominatim(user_agent='ycc')
location = geolocater.geocode(address)
latitude = location.latitude
longitude = location.longitude

# create map of Downtown Toronto
map_downtown = folium.Map(location=[latitude, longitude], zoom_start=13)

# add markers to map
for lat, lng, label in zip(downtown_df['Latitude'], downtown_df['Longitude'], downtown_df['Neighborhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='orange',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_downtown)  
    
map_downtown

III. --> define api access and functions to extract data for the neighborhoods of Downtown Toronto<br>
the credentials will be put to a hidden cell

In [16]:
# The code was removed by Watson Studio for sharing.

In [17]:
# compose url for api requests
VERSION = '20190101'
LIMIT = 100
radius = 50
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(CLIENT_ID, CLIENT_SECRET, VERSION, latitude, longitude, radius, LIMIT)

In [19]:
# a function to extract all neighborhoods' data of Downtown Toronto from foursquare, taken from the lab notebook DP0701EN-2-2-1
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

The above function will return a pandas data frame named nearby_venues, print the name of the neighborhoods before making repeated request to the foursquare api within a for-loop. 

In [20]:
# function call
downtown_venues = getNearbyVenues(names=downtown_df['Neighborhood'],
                                   latitudes=downtown_df['Latitude'],
                                   longitudes=downtown_df['Longitude']
                                  )

Rosedale
Cabbagetown, St. James Town
Church and Wellesley
Harbourfront, Regent Park
Ryerson, Garden District
St. James Town
Berczy Park
Central Bay Street
Adelaide, King, Richmond
Harbourfront East, Toronto Islands, Union Station
Design Exchange, Toronto Dominion Centre
Commerce Court, Victoria Hotel
Harbord, University of Toronto
Chinatown, Grange Park, Kensington Market
CN Tower, Bathurst Quay, Island airport, Harbourfront West, King and Spadina, Railway Lands, South Niagara
Stn A PO Boxes 25 The Esplanade
First Canadian Place, Underground city
Christie


check the data frame on its shape and features such as neighborhood/venues count and unique categories

In [21]:
# check the item number
print(downtown_venues.shape)
downtown_venues.head()

(1281, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0                     Rosedale              43.679563              -79.377529   
1                     Rosedale              43.679563              -79.377529   
2                     Rosedale              43.679563              -79.377529   
3                     Rosedale              43.679563              -79.377529   
4  Cabbagetown, St. James Town              43.667967              -79.367675   

                  Venue  Venue Latitude  Venue Longitude Venue Category  
0         Rosedale Park       43.682328       -79.378934     Playground  
1          Whitney Park       43.682036       -79.373788           Park  
2  Alex Murray Parkette       43.678300       -79.382773           Park  
3        Milkman's Lane       43.676352       -79.373842          Trail  
4           Cranberries       43.667843       -79.369407          Diner

In [22]:
# count how many venues every neighborhood has
downtown_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                                                                        
Adelaide, King, Richmond                                                                                                      100   
Berczy Park                                                                                                                    55   
CN Tower, Bathurst Quay, Island airport, Harbourfront West, King and Spadina, Railway Lands, South Niagara                     14   
Cabbagetown, St. James Town                                                                                                    42   
Central Bay Street                                                                                                             88   
Chinatown, Grange Park, Kensington Market                                                                                     100   
Christie                                                                                                                       16   
Church and Wellesley                                                                                                           88   
Commerce Court, Victoria Hotel                                                                                                100   
Design Exchange, Toronto Dominion Centre                                                                                      100   
First Canadian Place, Underground city                                                                                        100   
Harbord, University of Toronto                                                                                                 35   
Harbourfront East, Toronto Islands, Union Station                                                                             100   
Harbourfront, Regent Park                                                                                                      45   
Rosedale                                                                                                                        4   
Ryerson, Garden District                                                                                                      100   
St. James Town                                                                                                                100   
Stn A PO Boxes 25 The Esplanade                                                                                                94   

                                                                                                            Neighborhood Longitude  \
Neighborhood                                                                                                                         
Adelaide, King, Richmond                                                                                                       100   
Berczy Park                                                                                                                     55   
CN Tower, Bathurst Quay, Island airport, Harbourfront West, King and Spadina, Railway Lands, South Niagara                      14   
Cabbagetown, St. James Town                                                                                                     42   
Central Bay Street                                                                                                              88   
Chinatown, Grange Park, Kensington Market                                                                                      100   
Christie                                                                                                                        16   
Church and Wellesley                                                                                                            88   
Commerce Court, Victoria Hotel                                                                             

In [23]:
# check the number of unique categories
print('There are {} uniques categories.'.format(len(downtown_venues['Venue Category'].unique())))

There are 207 uniques categories.


IV. --> preparing data and cluster

In [24]:
# one hot encoding
downtown_onehot = pd.get_dummies(downtown_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
downtown_onehot['Neighborhood'] = downtown_venues['Neighborhood']

# move neighborhood column to the first column
fixed_columns = [downtown_onehot.columns[-1]] + list(downtown_onehot.columns[:-1])
downtown_onehot = downtown_onehot[fixed_columns]

downtown_onehot.head()

Yoga Studio  Adult Boutique  Afghan Restaurant  Airport  \
0            0               0                  0        0   
1            0               0                  0        0   
2            0               0                  0        0   
3            0               0                  0        0   
4            0               0                  0        0   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                   0             0               0                0   
1                   0             0               0                0   
2                   0             0               0                0   
3                   0             0               0                0   
4                   0             0               0                0   

   Airport Terminal  American Restaurant      ...        \
0                 0                    0      ...         
1                 0                    0      ...         
2                 0                    0      ...         
3                 0                    0      ...         
4                 0                    0      ...         

   Thrift / Vintage Store  Toy / Game Store  Trail  Train Station  \
0                       0                 0      0              0   
1                       0                 0      0              0   
2                       0                 0      0              0   
3                       0                 0      1              0   
4                       0                 0      0              0   

   Vegetarian / Vegan Restaurant  Video Game Store  Vietnamese Restaurant  \
0                              0                 0                      0   
1                              0                 0                      0   
2                              0                 0                      0   
3                              0                 0                      0   
4                              0                 0                      0   

   Wine Bar  Wings Joint  Women's Store  
0         0            0              0  
1         0            0              0  
2         0            0              0  
3         0            0              0  
4         0            0              0  

[5 rows x 207 columns]

In [25]:
# take mean (frequencies) for the venue types 
downtown_grouped = downtown_onehot.groupby('Neighborhood').mean().reset_index()
downtown_grouped

Neighborhood  \
0                                                                                     Adelaide, King, Richmond   
1                                                                                                  Berczy Park   
2   CN Tower, Bathurst Quay, Island airport, Harbourfront West, King and Spadina, Railway Lands, South Niagara   
3                                                                                  Cabbagetown, St. James Town   
4                                                                                           Central Bay Street   
5                                                                    Chinatown, Grange Park, Kensington Market   
6                                                                                                     Christie   
7                                                                                         Church and Wellesley   
8                                                                               Commerce Court, Victoria Hotel   
9                                                                     Design Exchange, Toronto Dominion Centre   
10                                                                      First Canadian Place, Underground city   
11                                                                              Harbord, University of Toronto   
12                                                           Harbourfront East, Toronto Islands, Union Station   
13                                                                                   Harbourfront, Regent Park   
14                                                                                                    Rosedale   
15                                                                                    Ryerson, Garden District   
16                                                                                              St. James Town   
17                                                                             Stn A PO Boxes 25 The Esplanade   

    Yoga Studio  Adult Boutique  Afghan Restaurant   Airport  \
0      0.000000        0.000000           0.000000  0.000000   
1      0.000000        0.000000           0.000000  0.000000   
2      0.000000        0.000000           0.000000  0.071429   
3      0.000000        0.000000           0.000000  0.000000   
4      0.011364        0.000000           0.000000  0.000000   
5      0.000000        0.000000           0.000000  0.000000   
6      0.000000        0.000000           0.000000  0.000000   
7      0.011364        0.011364           0.011364  0.000000   
8      0.000000        0.000000           0.000000  0.000000   
9      0.000000        0.000000           0.000000  0.000000   
10     0.000000        0.000000           0.000000  0.000000   
11     0.028571        0.000000           0.000000  0.000000   
12     0.000000        0.000000           0.000000  0.000000   
13     0.022222        0.000000           0.000000  0.000000   
14     0.000000        0.000000           0.000000  0.000000   
15     0.000000        0.000000           0.000000  0.000000   
16     0.000000        0.000000           0.000000  0.000000   
17     0.000000        0.000000           0.000000  0.000000   

    Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0             0.000000      0.000000        0.000000         0.000000   
1             0.000000      0.000000        0.000000         0.000000   
2             0.071429      0.071429        0.142857         0.214286   
3             0.000000      0.000000        0.000000         0.000000   
4             0.000000      0.000000        0.000000         0.000000   
5             0.000000      0.000000        0.000000         0.000000   
6             0.000000      0.000000        0.000000         0.000000   
7             0.000000      0.000000        0.000000         0.000000   
8             0.000000      0.000000        0.000000     

find the most common venues as indicator of the 'character' of every neighborhood

In [26]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [27]:
# find top 10 of every neighborhood
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = downtown_grouped['Neighborhood']

for ind in np.arange(downtown_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(downtown_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood  \
0                                                                                    Adelaide, King, Richmond   
1                                                                                                 Berczy Park   
2  CN Tower, Bathurst Quay, Island airport, Harbourfront West, King and Spadina, Railway Lands, South Niagara   
3                                                                                 Cabbagetown, St. James Town   
4                                                                                          Central Bay Street   

  1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
0           Coffee Shop                  Café   American Restaurant   
1           Coffee Shop          Cocktail Bar        Farmers Market   
2       Airport Service      Airport Terminal        Airport Lounge   
3           Coffee Shop                  Café           Pizza Place   
4           Coffee Shop                  Café    Italian Restaurant   

  4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
0                   Bar            Steakhouse       Thai Restaurant   
1    Italian Restaurant            Restaurant           Cheese Shop   
2       Harbor / Marina      Sculpture Garden               Airport   
3                Bakery            Restaurant                   Pub   
4          Burger Joint        Sandwich Place                   Spa   

  7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
0                 Hotel        Cosmetics Shop                Bakery   
1              Beer Bar    Seafood Restaurant                Bakery   
2    Airport Food Court          Airport Gate              Boutique   
3    Italian Restaurant                  Park         Deli / Bodega   
4     Indian Restaurant   Japanese Restaurant      Sushi Restaurant   

      10th Most Common Venue  
0               Burger Joint  
1                 Steakhouse  
2              Boat or Ferry  
3      General Entertainment  
4  Middle Eastern Restaurant

In [28]:
# cluster
kclusters = 5

downtown_grouped_clustering = downtown_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(downtown_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([1, 1, 0, 1, 1, 3, 4, 1, 1, 1], dtype=int32)

In [29]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

downtown_merged = downtown_df

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
downtown_merged = downtown_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

downtown_merged.head() # check the last columns!

PostalCode           Borough                 Neighborhood   Latitude  \
0        M4W  Downtown Toronto                     Rosedale  43.679563   
1        M4X  Downtown Toronto  Cabbagetown, St. James Town  43.667967   
2        M4Y  Downtown Toronto         Church and Wellesley  43.665860   
3        M5A  Downtown Toronto    Harbourfront, Regent Park  43.654260   
4        M5B  Downtown Toronto     Ryerson, Garden District  43.657162   

   Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0 -79.377529               2                  Park            Playground   
1 -79.367675               1           Coffee Shop                  Café   
2 -79.383160               1           Coffee Shop   Japanese Restaurant   
3 -79.360636               1           Coffee Shop                Bakery   
4 -79.378937               1           Coffee Shop        Clothing Store   

  3rd Most Common Venue 4th Most Common Venue        5th Most Common Venue  \
0                 Trail          Dance Studio  Eastern European Restaurant   
1           Pizza Place                Bakery                   Restaurant   
2      Sushi Restaurant            Restaurant                      Gay Bar   
3                  Park                   Pub           Mexican Restaurant   
4        Cosmetics Shop                  Café    Middle Eastern Restaurant   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0   Dumpling Restaurant            Donut Shop      Doner Restaurant   
1                   Pub    Italian Restaurant                  Park   
2           Men's Store                  Café                 Hotel   
3        Breakfast Spot                  Café               Theater   
4    Italian Restaurant              Tea Room      Ramen Restaurant   

  9th Most Common Venue 10th Most Common Venue  
0               Dog Run         Discount Store  
1         Deli / Bodega  General Entertainment  
2          Burger Joint        Bubble Tea Shop  
3  Gym / Fitness Center         Farmers Market  
4           Pizza Place    Sporting Goods Shop

V. --> finally plot the map with neighborhood clusters of Downtown Toronto

In [30]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=13)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(downtown_merged['Latitude'], downtown_merged['Longitude'], downtown_merged['Neighborhood'], downtown_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters